File: api/src/main/java/dev/obscuria/fragmentum/util/signal/Signal5.java
package dev.obscuria.fragmentum.util.signal;

import org.apache.commons.lang3.NotImplementedException;

@SuppressWarnings("unused")
public final class Signal5<P1, P2, P3, P4, P5> extends Signal<Signal5.Listener<P1, P2, P3, P4, P5>>
{
    public void emit(P1 p1, P2 p2, P3 p3, P4 p4, P5 p5)
    {
        throw new NotImplementedException();
    }

    @FunctionalInterface
    public interface Listener<P1, P2, P3, P4, P5>
    {
        void consume(P1 p1, P2 p2, P3 p3, P4 p4, P5 p5);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal5.kt
package dev.obscuria.fragmentum.tools.signal

class Signal5<P1, P2, P3, P4, P5> : Signal<Signal5.Listener<P1, P2, P3, P4, P5>>() {

    fun emit(p1: P1, p2: P2, p3: P3, p4: P4, p5: P5) = emit { it.consume(p1, p2, p3, p4, p5) }

    fun interface Listener<P1, P2, P3, P4, P5> {

        fun consume(p1: P1, p2: P2, p3: P3, p4: P4, p5: P5)
    }
}